### PR TITLE
added image dataset first implementation + many fixes

### PR DESCRIPTION
**Pull request by chanokin**
Hey, I hacked together a proposal for the Poisson image dataset Source IR class and the procedures to export.

I also:
 - Fixed some naming bugs on the ml_genn side of things 
 - Moved the default values for Connector parameter keys to the Constants package so I can test against these values when exporting Connections.
 - Added a name attribute to the Network object so we can use it to grab recordings back if populations have been chosen to record
 - Added the code generation procedures to export the fetching of recording results after the simulation has ended

### CODE REVIEW

File: bifrost/export/pynn.py [hunk 1]
```diff
@@ -38,4 +36,18 @@ class PyNNNeuronTypes(Enum):
...
     return Statement(f"Grid2D({ratio})",
-                     imports=['from pyNN.space import Grid2D'])
\ No newline at end of file
+                     imports=['from pyNN.space import Grid2D'])
+
+
+def export_constraints(constraints: Dict[str, Any]) -> Statement:
+    if 'max_neurons' in constraints:
+        return export_max_neurons_per_core(constraints['max_neurons'])
+
```
Comment (Jegp) [inline bifrost/export/pynn.py:45]: Would it make sense to raise an exception or log a warning if we don't support the constraint in the dict?

**Review by Jegp — COMMENTED** [review]
I understand the need for constraints, but currently, the implementation of constraints seems to be the same as with configurations. Is there a reason for keeping them separate? Can they be merged?

File: bifrost/export/pynn.py (hunk 1 above)
Comment (chanokin) [reply bifrost/export/pynn.py:45]: Yes to both, I don't know what I was thinking in splitting configs and constraints 😓  I'll make the changes

**Comment by Jegp** [comment]
This is excellent @chanokin! I really enjoy following your progress! We have a workshop coming up this weekend, but then I promise to spend some more time here. It's a great motivator to see you also find this appealing/worthwhile :)

I would suggest we just go ahead and merge this.

**Pull request merged**